File: 1110-delete-nodes-and-return-forest/1110-delete-nodes-and-return-forest.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode delNodes_01(TreeNode node,HashSet<Integer> map,ArrayList<TreeNode> ans) {
        if(node == null) return null;
        
        node.left = delNodes_01(node.left,map,ans);
        node.right = delNodes_01(node.right,map,ans);
        
        if(map.contains(node.val)) {
            if(node.left != null) {
                ans.add(node.left); 
            }
            
            if(node.right != null) {
                ans.add(node.right); 
            }
            
            return null;
        }
        
        return node;
    }
    
    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        HashSet<Integer> map = new HashSet<>();
        for(int ele : to_delete) {
            map.add(ele);
        }
        ArrayList<TreeNode> ans = new ArrayList<>();  
        if(!map.contains(root.val)) {
            ans.add(root);
        }
        delNodes_01(root,map,ans);
        return ans;
        
       
    }
}

File: 1110-delete-nodes-and-return-forest/1110-delete-nodes-and-return-forest.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* delNodes_01(TreeNode* node, vector<int>& to_delete,vector<TreeNode*>& ans) {
        if(!node) return nullptr; 
        //Recursive Calls
        node->left = delNodes_01(node->left,to_delete,ans);
        node->right = delNodes_01(node->right,to_delete,ans);
        
        //Deletion
        if(find(to_delete.begin(),to_delete.end(),node->val) != to_delete.end()) {
            if(node->left) {
                ans.push_back(node->left);
            }
            
            if(node->right) {
                ans.push_back(node->right);
            }
            return nullptr;
        }
        
        return node;
    }
    
    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete) {
        vector<TreeNode*> ans;
        
        if(delNodes_01(root,to_delete,ans)) {
            ans.push_back(root);
        }
        return ans;
    }
};